### PR TITLE
Use union to represent gameboy registers

### PR DESCRIPTION
**Pull request by marcluque**
The major feature is the change to use a union that represents a register of the gameboy.
Further changes are:
- Fix some stray clang warning messages
- Make include guards consistent

### CODE REVIEW

File: src/cpu.c [hunk 1]
```diff
@@ -229,367 +177,367 @@ void optable_init(void) {
...
-    mmu_write_byte(get_BC(), c.A);
-    c.cycle_count++;
+    mmu_write_byte(cpu.BC.word, cpu.AF.bytes.high);
+    cpu.cycle_count++;
 }
 
 void OPC_INC_BC(void) {
-    set_BC(get_BC()+1);
-    c.cycle_count += 2;
+    cpu.BC.word = cpu.BC.word+1;
```
Comment (marcluque) [inline src/cpu.c:246]:
```suggestion
    ++cpu.BC.word;
```

**Review by Benzammour — APPROVED** [review]

File: src/cpu.h [hunk 2]
```diff
@@ -1,55 +1,64 @@
...
 
-    uint8_t B;
-    uint8_t C;
+typedef union Register {
+	struct {
+		uint8_t high; // A, D, D, H
+		uint8_t low; // F, C, E, L
+	} bytes;
+	uint16_t word;
+} Register;
```
Comment (mormod) [inline src/cpu.h:31]: Is using upper case common standard for setting "types" in C?

File: src/cpu.h [hunk 3]
```diff
@@ -1,55 +1,64 @@
...
     uint16_t PC;
+
     uint16_t cycle_count;
+
     uint8_t opcode;
-} cpu;
+} Cpu;
 
-extern cpu c;
+extern Cpu cpu;
```
Comment (mormod) [inline src/cpu.h:47]: Same as above. This is much more an aesthetic thing then anything else, but `Cpu` just looks wrong to me.

File: src/cpu.c [hunk 4]
```diff
@@ -4,136 +4,84 @@
...
-}
-
-uint16_t get_AF(void) {
-    return (c.A << 8) | c.F;
-}
-
-void set_BC(uint16_t x) {
-    c.C = (uint8_t) x;
-    c.B = (uint8_t) (x >> 8);
+__attribute((always_inline)) inline static void set_flag(uint8_t bit, Flag f) {
```
Comment (mormod) [inline src/cpu.c:30]: What is the reasoning behind `always_inline` here? Shouldn't this be decided by the compiler (using the normal `inline` keyword as you've also done)?

File: src/cpu.c [hunk 5]
```diff
@@ -4,136 +4,84 @@
...
+	cpu.AF.bytes.low |= bit << f;
 }
 
-uint16_t get_BC(void) {
-    return (c.B << 8) | c.C;
+__attribute((always_inline)) inline static void clear_flag(uint8_t bit, Flag f) {
```
Comment (mormod) [inline src/cpu.c:34]: Same as above.

File: src/cpu.h (hunk 2 above)
Comment (marcluque) [reply src/cpu.h:31]: Google, LLVM, etc. do this, i.e., PascalCase
GNU is the only one doing snake_case

File: src/cpu.h (hunk 3 above)
Comment (marcluque) [reply src/cpu.h:47]: I think it's fine, CPU could be confused with UPPER_CASE such as macros.

File: src/cpu.h (hunk 2 above)
Comment (mormod) [reply src/cpu.h:31]: I see, thx

File: src/cpu.h (hunk 3 above)
Comment (mormod) [reply src/cpu.h:47]: It really is a shame that *_t is namespaced by POSIX.

File: src/cpu.c (hunk 4 above)
Comment (marcluque) [reply src/cpu.c:30]: I would argue that these one-liners should always be inlined even if you are doing a non-optimized build. So basically, it makes sure that the function is inlined even when optimizations are turned off.
Here are some "interesting" results from a benchmark:
image
https://indico.cern.ch/event/386232/sessions/159923/attachments/771039/1057534/always_inline_performance.pdf

File: src/cpu.h (hunk 3 above)
Comment (Benzammour) [reply src/cpu.h:47]: I would also argue for something else, it simply doesn't fit somehow :D
Any other suggestions?

Comment (marcluque) [reply src/cpu.h:47]: CentralProcessingUnit 😎

Comment (marcluque) [reply src/cpu.h:47]: I mean actually it won't break PascalCase if we use CPU 🤔 
I suppose I was a bit too quick before, it should be fine 🤔

File: src/cpu.c (hunk 4 above)
Comment (mormod) [reply src/cpu.c:30]: No doubt about that, So this is basically a little bit more advanced `inline`? But isn't `inline` taken into account without the need for optimization options?

Comment (marcluque) [reply src/cpu.c:30]: Advanced as in, we can force the compiler in almost all cases to inline? Then yes, maybe I'm misunderstanding something here?
For the latter question: In the scenario where we compile with -O0 oder -O1 inline has no effect anymore:

> -fno-inline
>     Do not expand any functions inline apart from those marked with the always_inline attribute. This is the default when not optimizing. Single functions can be exempted from inlining by marking them with the noinline attribute.

File: src/cpu.h (hunk 3 above)
Comment (mormod) [reply src/cpu.h]: PascalCase is fine for me if acronyms are respected and written in uppercase as they suppose to be

**Pull request merged**